File: test-case/validate/list2d/list2d.pl
use strict;
use warnings;
use List::Util qw(min max);
use Soda::DS::PriorityQueue;
use Soda::Leetcode::ListNode;
use Soda::Leetcode::NestedInteger;
use Soda::Leetcode::TreeNode;

*ListNode:: = \*Soda::Leetcode::ListNode::;
*NestedInteger:: = \*Soda::Leetcode::NestedInteger::;
*PriorityQueue:: = \*Soda::DS::PriorityQueue::;
*TreeNode:: = \*Soda::Unittest::TestWork::;

# step [1]: implement solution function
use POSIX qw/floor/;
use Data::Dumper;
# @param {string[]} strs
# @return {string[][]}
sub group_by_length {
    my $strs = shift;
    shuffle($strs);
    my %group;
    for my $s (@$strs) {
        $group{length($s)} = [] unless exists $group{length($s)};
        push @{$group{length($s)}}, $s;
    }
    my $keys = [keys %group];
    shuffle($keys);
    [map { $group{$_} } @$keys];
}

sub shuffle {
    my $arr = shift;
    for (my $i = @$arr; $i > 0; --$i) {
        my $index = floor(rand() * $i);
        if ($index != $i-1) {
            ($arr->[$index], $arr->[$i-1]) = ($arr->[$i-1], $arr->[$index]);
        }
    }
}

package main;
use Soda::Unittest::Utils qw(from_stdin);
use Soda::Unittest::TestWork qw(create for_struct);
use Soda::Unittest::Validators qw(for_array for_array2d);
# step [2]: setup function/return/arguments
my $work = create(\&group_by_length);
# my $work = for_struct(PACKAGE::);
$work->validator(for_array2d('string', 0, 0));
$work->compare_serial(1);
print $work->run(from_stdin());

File: test-case/validate/list2d/list2d.rb
# step [1]: implement solution function
# @param {String[]} strs
# @return {String[][]}
def group_by_length(strs)
  strs.shuffle!
  group = Hash.new{ |hash,key| hash[key] = Array.new }
  strs.each { |s|
    group[s.size] << s
  }
  group.keys.shuffle.map{ |k| group[k] }
end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  work = ns::TestWork.create(method(:group_by_length))
  # work = ns::TestWork.forStruct(CLASS)
  work.validator = -> (e, r) { ns::Validators.for_array2d('String', false, false) }
  work.compare_serial = true
  puts work.run(ARGF.read)
end

